### PR TITLE
GCMemcard: Rework construction logic.

### PR DESCRIPTION
**Pull request by AdmiralCurtiss**
This is to better match our knowledge of the format, while also providing better error reporting facilities. (no `PanicAlert()`s anymore!) Obsoletes #8020. Probably conflicts with #7996 and #7975 but that's okay, I'll just rebase as appropriate.

TODOs:

 - [x] Hardware test the Dir/BAT behavior. (see https://bugs.dolphin-emu.org/issues/11828)
 - [x] Implement cross-block error checks such as Dir/BAT data mismatch.
 - [x] Provide better error messages to the user in the Memcard manager and the Memcard settings when selecting an invalid card.

### CODE REVIEW

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 1]
```diff
@@ -1473,6 +1483,27 @@ std::pair<u16, u16> Header::CalculateChecksums() const
...
+  GCMemcardErrorCode error_code;
+
+  // total card size should match card size in header
+  if (m_size_mb != card_size_mbits)
+    error_code.Set(GCMemcardValidityIssues::INVALID_CARD_SIZE);
+
+  // unused areas, should always be filled with 0xFF
+  if (std::any_of(m_unused_1.begin(), m_unused_1.end(), [](u8 val) { return val != 0xFF; }) ||
+      std::any_of(m_unused_2.begin(), m_unused_2.end(), [](u8 val) { return val != 0xFF; }))
+    error_code.Set(GCMemcardValidityIssues::DATA_IN_UNUSED_AREA);
```
Comment (lioncash) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: This should be braced, since the condition travels more than one line.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 2]
```diff
@@ -679,6 +652,43 @@ std::pair<u16, u16> BlockAlloc::CalculateChecksums() const
...
+  if (checksum_sum != m_checksum || checksum_inv != m_checksum_inv)
+    error_code.Set(GCMemcardValidityIssues::INVALID_CHECKSUM);
+
+  if (size_mbits > 0 && size_mbits <= 256)
+  {
+    // check if free block count matches the actual amount of free blocks in m_map
+    u16 total_available_blocks = (size_mbits * MBIT_TO_BLOCKS) - MC_FST_BLOCKS;
+    assert(total_available_blocks <= m_map.size());
+    u16 blocks_in_use = 0;
+    for (size_t i = 0; i < total_available_blocks; ++i)
```
Comment (lioncash) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: This should be braced

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 3]
```diff
@@ -679,6 +652,43 @@ std::pair<u16, u16> BlockAlloc::CalculateChecksums() const
...
+      if (m_map[i] != 0)
+        ++blocks_in_use;
+    u16 free_blocks = total_available_blocks - blocks_in_use;
+
+    if (free_blocks != m_free_blocks)
+      error_code.Set(GCMemcardValidityIssues::FREE_BLOCK_MISMATCH);
+
+    // remaining blocks map to nothing on hardware and must be empty
+    for (size_t i = total_available_blocks; i < m_map.size(); ++i)
```
Comment (lioncash) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Ditto

File: Source/Core/Core/HW/GCMemcard/GCMemcard.h [hunk 4]
```diff
@@ -79,6 +80,29 @@ enum class GCMemcardRemoveFileRetVal
...
+  INVALID_CARD_SIZE,
+  INVALID_CHECKSUM,
+  FREE_BLOCK_MISMATCH,
+  DATA_IN_UNUSED_AREA,
+  COUNT
+};
+
+class GCMemcardErrorCode
+{
+private:
```
Comment (lioncash) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.h]: Public interface should come before the private interface. Existing instances where this isn't the case is due to code existing before the coding guideline was introduced.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.h [hunk 5]
```diff
@@ -79,6 +80,29 @@ enum class GCMemcardRemoveFileRetVal
...
+  std::bitset<static_cast<size_t>(GCMemcardValidityIssues::COUNT)> errors;
```
Comment (lioncash) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.h]:
```suggestion
  std::bitset<static_cast<size_t>(GCMemcardValidityIssues::COUNT)> m_errors;
```

We use the `m_` prefix for private class variables.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 6]
```diff
@@ -27,183 +27,181 @@ static void ByteSwap(u8* valueA, u8* valueB)
...
 
-  if (!mcdFile.ReadBytes(&m_bat_blocks[1], BLOCK_SIZE))
+  // check if the filesize is a valid memory card size
+  u16 card_size_mbits;
+  u64 filesize = file.GetSize();
+  switch (filesize)
   {
-    PanicAlertT("Failed to read 2nd block allocation table block correctly\n(0x8000-0x9FFF)");
-    return;
+  case (MemCard59Mb * MBIT_TO_BLOCKS) * BLOCK_SIZE:
```
Comment (Warepire) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: What do you think of a small constexpr function to generate this math?

Example:
```suggestion
// Somewhere else:
constexpr u64 MemCardSizeToFileSize(u16 card_size) {
  return card_size * MBIT_TO_BLOCKS * BLOCK_SIZE;
}

// case label:
  case MemCardSizeToFileSize(MemCard59Mb):
```

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 7]
```diff
@@ -27,183 +27,181 @@ static void ByteSwap(u8* valueA, u8* valueB)
...
+    return std::make_pair(error_code, std::nullopt);
+
+  // can return invalid card size, invalid checksum, data in unused area, free block mismatch
+  // if both bats are invalid: fail
+  // if only one is invalid: copy the other one over and continue
+  GCMemcardErrorCode bat_block_0_error_code = card.m_bat_blocks[0].CheckForErrors(card_size_mbits);
+  GCMemcardErrorCode bat_block_1_error_code = card.m_bat_blocks[1].CheckForErrors(card_size_mbits);
+
+  bool bat_block_0_valid = !bat_block_0_error_code.HasCriticalErrors();
+  bool bat_block_1_valid = !bat_block_1_error_code.HasCriticalErrors();
```
Comment (Warepire) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]:
```suggestion
  const bool bat_block_1_valid = !bat_block_1_error_code.HasCriticalErrors();
```

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 8]
```diff
@@ -27,183 +27,181 @@ static void ByteSwap(u8* valueA, u8* valueB)
...
+  if (!dir_block_0_valid && !dir_block_1_valid)
+    return std::make_pair(error_code, std::nullopt);
+
+  // can return invalid card size, invalid checksum, data in unused area, free block mismatch
+  // if both bats are invalid: fail
+  // if only one is invalid: copy the other one over and continue
+  GCMemcardErrorCode bat_block_0_error_code = card.m_bat_blocks[0].CheckForErrors(card_size_mbits);
+  GCMemcardErrorCode bat_block_1_error_code = card.m_bat_blocks[1].CheckForErrors(card_size_mbits);
+
+  bool bat_block_0_valid = !bat_block_0_error_code.HasCriticalErrors();
```
Comment (Warepire) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]:
```suggestion
  const bool bat_block_0_valid = !bat_block_0_error_code.HasCriticalErrors();
```

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 9]
```diff
@@ -27,183 +27,181 @@ static void ByteSwap(u8* valueA, u8* valueB)
...
+  // one, but this is safer in case eg. FixChecksums() is called later and shouldn't make a
+  // practical difference.
+  // Also while we're at it test what happens on free block mismatch.
+
+  // can return invalid checksum, data in unused area
+  // if both checksums are invalid: fail
+  // if only one is invalid: copy the valid block over the invalid one and continue
+  GCMemcardErrorCode dir_block_0_error_code = card.m_directory_blocks[0].CheckForErrors();
+  GCMemcardErrorCode dir_block_1_error_code = card.m_directory_blocks[1].CheckForErrors();
+  bool dir_block_0_valid = !dir_block_0_error_code.HasCriticalErrors();
```
Comment (Warepire) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]:
```suggestion
  const bool dir_block_0_valid = !dir_block_0_error_code.HasCriticalErrors();
```

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 10]
```diff
@@ -27,183 +27,181 @@ static void ByteSwap(u8* valueA, u8* valueB)
...
+  bool dir_block_1_valid = !dir_block_1_error_code.HasCriticalErrors();
```
Comment (Warepire) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]:
```suggestion
  const bool dir_block_1_valid = !dir_block_1_error_code.HasCriticalErrors();
```

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 11]
```diff
@@ -27,183 +27,181 @@ static void ByteSwap(u8* valueA, u8* valueB)
...
-    // header checksum error!
-    // invalid files do not always get here
-    PanicAlertT("Header checksum failed");
-    return;
+    error_code.Set(GCMemcardValidityIssues::IO_ERROR);
+    return std::make_pair(error_code, std::nullopt);
   }
 
-  if (csums & 0x2)  // 1st directory block checksum error!
+  u16 card_size_blocks = card_size_mbits * MBIT_TO_BLOCKS;
```
Comment (Warepire) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]:
```suggestion
  const u16 card_size_blocks = card_size_mbits * MBIT_TO_BLOCKS;
```

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 12]
```diff
@@ -27,183 +27,181 @@ static void ByteSwap(u8* valueA, u8* valueB)
...
+  u16 user_data_blocks = card_size_blocks - MC_FST_BLOCKS;
```
Comment (Warepire) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]:
```suggestion
  const u16 user_data_blocks = card_size_blocks - MC_FST_BLOCKS;
```

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 13]
```diff
@@ -27,183 +27,181 @@ static void ByteSwap(u8* valueA, u8* valueB)
...
-    PanicAlertT("Failed to read 1st block allocation table block correctly\n(0x6000-0x7FFF)");
-    return;
+    error_code.Set(GCMemcardValidityIssues::FAILED_TO_OPEN);
+    return std::make_pair(error_code, std::nullopt);
   }
 
-  if (!mcdFile.ReadBytes(&m_bat_blocks[1], BLOCK_SIZE))
+  // check if the filesize is a valid memory card size
+  u16 card_size_mbits;
+  u64 filesize = file.GetSize();
```
Comment (Warepire) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]:
```suggestion
  const u64 filesize = file.GetSize();
```

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp (hunk 6 above)
Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: I made a more generic `MegabitsToBytes()` instead but yeah, that's reasonable.

**Comment by AdmiralCurtiss** [comment]
Okay, I think I'm done with this now. Please review again.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 14]
```diff
@@ -27,196 +27,256 @@ static void ByteSwap(u8* valueA, u8* valueB)
...
-      break;
-    default:
-      PanicAlertT("%s failed to load as a memory card.\nCard size is invalid (0x%x bytes)",
-                  filename.c_str(), (unsigned)size);
-      return;
-    }
-  }
+bool GCMemcardErrorCode::HasCriticalErrors() const
+{
+  return Test(GCMemcardValidityIssues::FAILED_TO_OPEN) || Test(GCMemcardValidityIssues::IO_ERROR) ||
```
Comment (BhaaLseN) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: I wonder if this is nicer as (warning, half pseudo-code):
```c++
std::array<GCMemcardValidityIssues, 7> critical_errors {{ FAILED_TO_OPEN, IO_ERROR, ... }};
return std::any_of(critical_errors, Test);
```

I briefly considered suggesting checking whether _only_ non-critical errors are present (since theres only one of those, vs. 7 critical ones); but for clarity this seems the nicer (albeit longer) way.

It's also sad that `std::bitset` has no real mass-test (whether any given number of bits are set), only those that test for all/any/none and specific bits. There is bitwise operators, but I don't think this should be rewritten as `(m_errors & critical_error_flags) != 0` (again for clariry/readabilty)

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 15]
```diff
@@ -27,196 +27,256 @@ static void ByteSwap(u8* valueA, u8* valueB)
...
-    return;
+    error_code.Set(GCMemcardValidityIssues::FAILED_TO_OPEN);
+    return std::make_pair(error_code, std::nullopt);
   }
 
-  if (!mcdFile.ReadBytes(&m_bat_blocks[1], BLOCK_SIZE))
+  // check if the filesize is a valid memory card size
+  u16 card_size_mbits;
+  const u64 filesize = file.GetSize();
+  switch (filesize)
```
Comment (BhaaLseN) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: This feels verbose and repeative...could we instead do `BytesToMegabits(filesize)` and just compare it to the list of known-good memory card sizes instead?

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 16]
```diff
@@ -27,196 +27,256 @@ static void ByteSwap(u8* valueA, u8* valueB)
...
   }
 
-  if (csums & 0x2)  // 1st directory block checksum error!
+  const u16 card_size_blocks = card_size_mbits * MBIT_TO_BLOCKS;
+  const u16 user_data_blocks = card_size_blocks - MC_FST_BLOCKS;
+  card.m_data_blocks.reserve(user_data_blocks);
+  for (u16 i = 0; i < user_data_blocks; ++i)
   {
-    if (csums & 0x4)
+    card.m_data_blocks.emplace_back();
```
Comment (BhaaLseN) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Since C++17 this returns the inserted element; can we just use it for clarity?
```c++
auto& new_data_block = card.m_data_blocks.emplace_back();
if (!file.ReadArray(new_data_block.m_block.data(), BLOCK_SIZE))
```
Or are there any compilers that still don't fully support C++17?

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 17]
```diff
@@ -27,196 +27,256 @@ static void ByteSwap(u8* valueA, u8* valueB)
...
+  if (!dir_block_0_valid)
+    ++number_of_corrupted_dir_bat_blocks;
+  if (!dir_block_1_valid)
+    ++number_of_corrupted_dir_bat_blocks;
+  if (!bat_block_0_valid)
+    ++number_of_corrupted_dir_bat_blocks;
+  if (!bat_block_1_valid)
+    ++number_of_corrupted_dir_bat_blocks;
+
+  if (number_of_corrupted_dir_bat_blocks > 1)
```
Comment (BhaaLseN) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: I wonder if we should allow this as potential data recovery option (when one dir/bat pair is still valid but the GC BIOS would reject it) - but this is probably out of scope here.

File: Source/Core/Core/HW/GCMemcard/GCMemcardDirectory.cpp [hunk 18]
```diff
@@ -703,12 +703,12 @@ void MigrateFromMemcardFile(const std::string& directory_name, int card_index)
                                                 Config::Get(Config::MAIN_MEMCARD_B_PATH);
   if (File::Exists(ini_memcard))
   {
-    GCMemcard memcard(ini_memcard.c_str());
-    if (memcard.IsValid())
+    auto memcard = GCMemcard::Open(ini_memcard.c_str());
```
Comment (BhaaLseN) [inline Source/Core/Core/HW/GCMemcard/GCMemcardDirectory.cpp]: How about using structured bindings instead of `.first`/`.second`?
```suggestion
    auto [error_code, memcard] = GCMemcard::Open(ini_memcard.c_str());
```

File: Source/Core/DolphinQt/GCMemcardManager.cpp [hunk 19]
```diff
@@ -241,13 +243,20 @@ void GCMemcardManager::UpdateActions()
 
 void GCMemcardManager::SetSlotFile(int slot, QString path)
 {
-  auto memcard = std::make_unique<GCMemcard>(path.toStdString());
+  auto memcard = GCMemcard::Open(path.toStdString());
```
Comment (BhaaLseN) [inline Source/Core/DolphinQt/GCMemcardManager.cpp]:
```suggestion
  auto [error_code, memcard] = GCMemcard::Open(path.toStdString());
```

File: Source/Core/DolphinQt/Settings/GameCubePane.cpp [hunk 20]
```diff
@@ -212,15 +213,14 @@ void GameCubePane::OnConfigPressed(int slot)
   {
     if (File::Exists(filename.toStdString()))
     {
-      GCMemcard mc(filename.toStdString());
-
-      if (!mc.IsValid())
+      auto mc = GCMemcard::Open(filename.toStdString());
```
Comment (BhaaLseN) [inline Source/Core/DolphinQt/Settings/GameCubePane.cpp]:
```suggestion
      auto [error_code, memcard] = GCMemcard::Open(filename.toStdString());
```

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp (hunk 14 above)
Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: I tried this and ended up with:

~~~
  std::array<GCMemcardValidityIssues, 7> critical_errors{
      {GCMemcardValidityIssues::FAILED_TO_OPEN, GCMemcardValidityIssues::IO_ERROR,
       GCMemcardValidityIssues::INVALID_CARD_SIZE, GCMemcardValidityIssues::INVALID_CHECKSUM,
       GCMemcardValidityIssues::MISMATCHED_CARD_SIZE, GCMemcardValidityIssues::FREE_BLOCK_MISMATCH,
       GCMemcardValidityIssues::DIR_BAT_INCONSISTENT}};
  return std::any_of(critical_errors.begin(), critical_errors.end(),
                     [&](GCMemcardValidityIssues e) { return Test(e); });
~~~

I'm not sure if that's really any better, to be honest.

Comment (BhaaLseN) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: IIRC a trailing comma after the last one lets lint be more lenient and allow one per line.
But I guess I'm ok with either solution at this point, since all of them look kinda ugly in the end.

Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: I'll just leave it as-is then.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp (hunk 17 above)
Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: I left a note in the code about potentially adding that later.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp (hunk 16 above)
Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: I didn't realize that was a thing! Yeah, that's definitely nicer.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp (hunk 15 above)
Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Like this?
